PreparedSQLPlaceholders: massive improvements

jrfnl opened the pull request:
I initially wanted to fine-tune the sniff to throw more helpful messages for `%` SQL wildcards found in the value part of a `LIKE` query based on https://core.trac.wordpress.org/changeset/41660 .
It took some time to figure out the right way to isolate those out correctly.

While working on that, I made a number of significant other improvements to the sniff making it far more accurate and helpful.

Please see the individual commits for more details.
The PR will also be easiest to review using the individual commits.

Extensive unit tests have been added for all additional checks and improvements.

----

### Todo once this has been merged:
- [ ] Update the wiki page for the new whitelist flag

JDGrimes commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -23,6 +23,14 @@
  *
+ * The sniff allows for a specific pattern with a variable number of placeholders
+ * created using code along the lines of:
+ * `sprintf( 'query .... IN (%s) ...', implode( ',', array_fill( 0, count( $something), '%s' ) ) )`.
```
There is a space missing after `$something`.

jrfnl replied:
Good point. Fixed.

For the example I agree that the space should be there. As for the sniff: that should work independently of code-style.

JDGrimes commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -217,4 +439,128 @@ public function process_token( $stackPtr ) {
+	 * to deal correctly with `IN` queries.
+	 *
+	 * The pattern we are searching for is:
+	 * `sprintf( 'query ....', implode( ',', array_fill( 0, count( $something), '%s' ) ) )`
```
Ditto.

JDGrimes commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -217,4 +439,128 @@ public function process_token( $stackPtr ) {
+	 * to dynamically create placeholders.
+	 *
+	 * The pattern we are searching for is:
+	 * `implode( ',', array_fill( 0, count( $something), '%s' ) )`
```
And here too.

jrfnl commented:
@JDGrimes While I fixed up those docs, I realized I should make one more tiny message precision improvement and add something about `LIKE` to the class docblock, so I've added both these changes in a separate new commit to make them reviewable rather than fixing up the original commits.

GaryJones commented on WordPress/Sniff.php:
```diff
@@ -36,6 +36,15 @@
+	 *
+	 * @var string
+	 */
+	const REGEX_COMPLEX_VARS = '`(?:(\{)?(?<!\\\\)\$)?(\{)?(?<!\\\\)\$(\{)?(?P<varname>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)(?:->\$?(?P>varname)|\[[^\]]+\]|::\$?(?P>varname)|\([^\)]*\))*(?(3)\}|)(?(2)\}|)(?(1)\}|)`';
```
Where has this regex come from? 

Can it be broken down and commented (with the flag that ignores whitespace)?

jrfnl replied:
I wrote the regex (what did you expect ?).

I did so nearly a year ago when I needed the same `strip_interpolated_variables()` function for the [PHPCompatibility standard](https://github.com/wimg/PHPCompatibility/blob/9a9d15aa50b6ffec7a0e7683377bef338c64911f/PHPCompatibility/Sniff.php#L26) and pulled it in https://github.com/wimg/PHPCompatibility/pull/314 which includes [extensive unit tests of the function](https://github.com/wimg/PHPCompatibility/blob/9a9d15aa50b6ffec7a0e7683377bef338c64911f/PHPCompatibility/Tests/BaseClass/FunctionsTest.php#L441-L504).

The varname part of the regex comes from http://php.net/manual/en/language.variables.basics.php

I don't mind documenting the regex, but that will easily take me two hours or more as I'll have to decipher it myself again first... 

If you insist on it being documented, I'd suggest opening an issue so it will be done in a separate PR at some point.
